File: quera-problems/series4/java/src/XYZ.java
import java.util.Scanner;

public class XYZ {
  public static void main(String[] args) {
    Scanner input = new Scanner(System.in);
    int arrayLength = input.nextInt();
    int[][] cards = new int[arrayLength][];
    for (int i = 0; i < arrayLength; i++) {
      String cardsString = input.next();
      switch (cardsString){
        case "X1": cards[i] = new int[]{0, 0}; break;
        case "X2": cards[i] = new int[]{0, 1}; break;
        case "X3": cards[i] = new int[]{0, 2}; break;
        case "X4": cards[i] = new int[]{0, 3}; break;
        case "X5": cards[i] = new int[]{0, 4}; break;
        case "Y1": cards[i] = new int[]{1, 0}; break;
        case "Y2": cards[i] = new int[]{1, 1}; break;
        case "Y3": cards[i] = new int[]{1, 2}; break;
        case "Y4": cards[i] = new int[]{1, 3}; break;
        case "Y5": cards[i] = new int[]{1, 4}; break;
        case "Z1": cards[i] = new int[]{2, 0}; break;
        case "Z2": cards[i] = new int[]{2, 1}; break;
        case "Z3": cards[i] = new int[]{2, 2}; break;
        case "Z4": cards[i] = new int[]{2, 3}; break;
        case "Z5": cards[i] = new int[]{2, 4}; break;
      }
    }
    cards = radixSort(cards);
    for (int i = 0; i < arrayLength; i++) {
      int numberRepresentation = (cards[i][0] * 10) + cards[i][1];
      switch (numberRepresentation){
        case 0: System.out.print("X1"); break;
        case 1: System.out.print("X2"); break;
        case 2: System.out.print("X3"); break;
        case 3: System.out.print("X4"); break;
        case 4: System.out.print("X5"); break;
        case 10: System.out.print("Y1"); break;
        case 11: System.out.print("Y2"); break;
        case 12: System.out.print("Y3"); break;
        case 13: System.out.print("Y4"); break;
        case 14: System.out.print("Y5"); break;
        case 20: System.out.print("Z1"); break;
        case 21: System.out.print("Z2"); break;
        case 22: System.out.print("Z3"); break;
        case 23: System.out.print("Z4"); break;
        case 24: System.out.print("Z5"); break;
      }
      System.out.print(" ");
    }
  }

  private static int[][] radixSort(int[][] cards){
    int[][] outputCards = cards;
    for (int i = 1; i >= 0; i--) {
      outputCards = countingSort(outputCards,i,5);
    }
    return outputCards;
  }

  private static int[][] countingSort(int[][] inputArray,int index,int range){
    int[][] outputArray = new int[inputArray.length][];
    int[] cumulativeFrequency = new int[range];
    for (int i = 0; i < inputArray.length; i++) {
      cumulativeFrequency[inputArray[i][index]]++;
    }
    for (int i = 1; i < cumulativeFrequency.length; i++){
      cumulativeFrequency[i] += cumulativeFrequency[i - 1];
    }
    for (int i = inputArray.length - 1; i >= 0; i--) {
      outputArray[cumulativeFrequency[inputArray[i][index]] - 1] = inputArray[i];
      cumulativeFrequency[inputArray[i][index]]--;
    }
    return outputArray;
  }
}

File: quera-problems/series4/cpp/xyz.cpp
#include <iostream>

using namespace std;


void countingSort(int inputArray[][2],int index,int range,int arrayLength){
    int outputArray[arrayLength][2];
    int cumulativeFrequency[range + 1];

    for (int i = 0; i <= range; i++) {
        cumulativeFrequency[i] = 0;
    }
    for (int i = 0; i < arrayLength; i++) {
        cumulativeFrequency[inputArray[i][index]]++;
    }
    for (int i = 1; i < range; i++){
        cumulativeFrequency[i] += cumulativeFrequency[i - 1];
    }
    for (int i = arrayLength - 1; i >= 0; i--) {
        outputArray[cumulativeFrequency[inputArray[i][index]] - 1][0] = inputArray[i][0];
        outputArray[cumulativeFrequency[inputArray[i][index]] - 1][1] = inputArray[i][1];
        cumulativeFrequency[inputArray[i][index]]--;
    }

    for (int i = 0; i < arrayLength; ++i){
        inputArray[i][0] = outputArray[i][0];
        inputArray[i][1] = outputArray[i][1];
    }
}

void radixSort(int cards[][2],int array_length){
    countingSort(cards,1,5,array_length);
    countingSort(cards,0,5,array_length);
}


int main(){
    int array_length;
    cin >> array_length;
    int cards[array_length][2];
    for (int i = 0; i < array_length; i++){
        string card_string = "";
        cin >> card_string;
        if ("X1" == card_string){
            cards[i][0] = 0;
            cards[i][1] = 0;
        } else if ("X2" == card_string){
            cards[i][0] = 0;
            cards[i][1] = 1;
        } else if ("X3" == card_string){
            cards[i][0] = 0;
            cards[i][1] = 2;
        } else if ("X4" == card_string){
            cards[i][0] = 0;
            cards[i][1] = 3;
        } else if ("X5" == card_string){
            cards[i][0] = 0;
            cards[i][1] = 4;
        } else if ("Y1" == card_string){
            cards[i][0] = 1;
            cards[i][1] = 0;
        } else if ("Y2" == card_string){
            cards[i][0] = 1;
            cards[i][1] = 1;
        } else if ("Y3" == card_string){
            cards[i][0] = 1;
            cards[i][1] = 2;
        } else if ("Y4" == card_string){
            cards[i][0] = 1;
            cards[i][1] = 3;
        } else if ("Y5" == card_string){
            cards[i][0] = 1;
            cards[i][1] = 4;
        } else if ("Z1" == card_string){
            cards[i][0] = 2;
            cards[i][1] = 0;
        } else if ("Z2" == card_string){
            cards[i][0] = 2;
            cards[i][1] = 1;
        } else if ("Z3" == card_string){
            cards[i][0] = 2;
            cards[i][1] = 2;
        } else if ("Z4" == card_string){
            cards[i][0] = 2;
            cards[i][1] = 3;
        } else if ("Z5" == card_string){
            cards[i][0] = 2;
            cards[i][1] = 4;
        }
    }
    
    radixSort(cards,array_length);

    for (int i = 0; i < array_length; i++) {
      int numberRepresentation = (cards[i][0] * 10) + cards[i][1];
      switch (numberRepresentation){
        case 0: cout << "X1"; break;
        case 1: cout << "X2"; break;
        case 2: cout << "X3"; break;
        case 3: cout << "X4"; break;
        case 4: cout << "X5"; break;
        case 10: cout << "Y1"; break;
        case 11: cout << "Y2"; break;
        case 12: cout << "Y3"; break;
        case 13: cout << "Y4"; break;
        case 14: cout << "Y5"; break;
        case 20: cout << "Z1"; break;
        case 21: cout << "Z2"; break;
        case 22: cout << "Z3"; break;
        case 23: cout << "Z4"; break;
        case 24: cout << "Z5"; break;
      }
      cout << " ";
    }
    return 0;
}